File: 0003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.go
func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
func lengthOfLongestSubstring(s string) int {
	s_len := len(s)
	if s_len <= 1 {
		return s_len
	}

	var (
		head int = 0
		ret  int = 0
	)
	m := map[rune]int{}

	for tail, c := range s {
		if index, ok := m[c]; ok && index != 0 {
			ret = Max(tail-head, ret)
			head = Max(head, index)

			if s_len-head < ret {
				return ret
			}
		}
		m[c] = tail + 1
	}
	return Max(s_len-head, ret)
}

File: 0003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.php
<?php
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        $s_len = strlen($s);
        if ($s_len <= 1) {
            return $s_len;
        }

        $head = 0;
        $ret = 0;
        $map = [];

        for ($index = 0; $index < $s_len;) {
            if (isset($map[$s[$index]])) {
                $ret = max($index - $head, $ret);
                $head = max($map[$s[$index]], $head);

                if ($s_len - $head < $ret) {
                    return $ret;
                }
            }
            $map[$s[$index]] = ++$index;
        }
        return max($s_len - $head, $ret);
    }
}
?>
